File: frontend/src/api/features.js
import api from './client'
import mockRequest, { MOCKS_ENABLED } from './mock/featuresMock'

export { MOCKS_ENABLED }
export { getApiError } from './finance'

// 신규 기능 API (backend.zip 의 routes/features.py 기준으로 경로 확정).
// 백엔드가 아직 Flask 에 등록되지 않아, VITE_USE_MOCKS=true 이면 목 응답을 돌려준다.
async function call(method, url, { params, data } = {}) {
  if (MOCKS_ENABLED) return mockRequest(method, url, { params, data })
  const response = await api.request({ method, url, params, data })
  return response.data
}

// multipart/form-data 업로드 (이미지). axios 인스턴스의 기본 JSON 헤더를 비운다.
async function upload(url, files) {
  if (MOCKS_ENABLED) return mockRequest('post', url, { data: { _files: files.length } })
  const form = new FormData()
  Array.from(files).forEach((file) => form.append('images', file))
  const response = await api.post(url, form, { headers: { 'Content-Type': undefined } })
  return response.data
}

// 대시보드
export const getDashboard = () => call('get', '/dashboard')
export const getDashboardHistory = (params = {}) => call('get', '/dashboard/history', { params })

// 미래 자산 시뮬레이션
export const runFreeProjection = (payload) => call('post', '/simulations/free', { data: payload })
export const runGoalProjection = (goalId, payload) => call('post', `/goals/${goalId}/simulation`, { data: payload })

// 저축 목표
export const getGoals = () => call('get', '/goals')
export const getGoal = (goalId) => call('get', `/goals/${goalId}`)
export const createGoal = (payload) => call('post', '/goals', { data: payload })
export const updateGoal = (goalId, payload) => call('patch', `/goals/${goalId}`, { data: payload })
export const deleteGoal = (goalId) => call('delete', `/goals/${goalId}`)

// 거래 내역
export const getLedgerTransactions = (params = {}) => call('get', '/transactions', { params })
export const getLedgerTransaction = (transactionId) => call('get', `/transactions/${transactionId}`)
export const getInvestmentTransactions = (params = {}) => call('get', '/investments/transactions', { params })

// 뱃지
export const getBadgeCatalog = () => call('get', '/badges')
export const getMyBadges = () => call('get', '/badges/me')

// 커뮤니티
export const getPosts = (params = {}) => call('get', '/posts', { params })
export const getPost = (postId) => call('get', `/posts/${postId}`)
export const createPost = (payload) => call('post', '/posts', { data: payload })
export const updatePost = (postId, payload) => call('patch', `/posts/${postId}`, { data: payload })
export const deletePost = (postId) => call('delete', `/posts/${postId}`)
export const getComments = (postId, params = {}) => call('get', `/posts/${postId}/comments`, { params })
export const createComment = (postId, payload) => call('post', `/posts/${postId}/comments`, { data: payload })
export const updateComment = (commentId, payload) => call('patch', `/comments/${commentId}`, { data: payload })
export const deleteComment = (commentId) => call('delete', `/comments/${commentId}`)
export const reactToPost = (postId, reactionType) => call('put', `/posts/${postId}/reaction`, { data: { reaction_type: reactionType } })
export const reportContent = (payload) => call('post', '/reports', { data: payload })
export const getMyReports = (params = {}) => call('get', '/reports/me', { params })
export const uploadPostImages = (postId, files) => upload(`/posts/${postId}/images`, files)
export const deleteAttachment = (attachmentId) => call('delete', `/attachments/${attachmentId}`)

// 거래 오류 문의
export const getInquiries = (params = {}) => call('get', '/inquiries', { params })
export const getInquiry = (inquiryId) => call('get', `/inquiries/${inquiryId}`)
export const createInquiry = (payload) => call('post', '/inquiries', { data: payload })
export const uploadInquiryImages = (inquiryId, files) => upload(`/inquiries/${inquiryId}/images`, files)

// 프로필
export const getMyProfile = () => call('get', '/profiles/me')
export const getProfile = (userId) => call('get', `/profiles/${userId}`)
export const updateMyProfile = (payload) => call('patch', '/profiles/me', { data: payload })
export const getMyVisibility = () => call('get', '/profiles/me/visibility')
export const updateMyVisibility = (payload) => call('patch', '/profiles/me/visibility', { data: payload })

// 관리자 (/api/admin/*, ADMIN 권한 필요; 비관리자는 403)
export const adminGetUsers = (params = {}) => call('get', '/admin/users', { params })
export const adminGetUser = (userId) => call('get', `/admin/users/${userId}`)
export const adminSetUserStatus = (userId, payload) => call('patch', `/admin/users/${userId}/status`, { data: payload })
export const adminDeleteUser = (userId, reason) => call('delete', `/admin/users/${userId}`, { data: { reason } })
export const adminAdjustAccount = (userId, payload) => call('post', `/admin/users/${userId}/adjustments`, { data: payload })
export const adminGetProducts = (params = {}) => call('get', '/admin/products', { params })
export const adminGetProduct = (productId) => call('get', `/admin/products/${productId}`)
export const adminPatchProduct = (productId, payload) => call('patch', `/admin/products/${productId}`, { data: payload })
export const adminPatchOption = (productId, optionId, payload) => call('patch', `/admin/products/${productId}/options/${optionId}`, { data: payload })
export const adminDeleteProduct = (productId, reason) => call('delete', `/admin/products/${productId}`, { data: { reason } })
export const adminGetLedger = (params = {}) => call('get', '/admin/transactions', { params })
export const adminGetMarketTransactions = (params = {}) => call('get', '/admin/market-transactions', { params })
export const adminGetPosts = (params = {}) => call('get', '/admin/posts', { params })
export const adminGetCommentsAll = (params = {}) => call('get', '/admin/comments', { params })
export const adminDeletePost = (postId, reason) => call('delete', `/admin/posts/${postId}`, { data: { reason } })
export const adminDeleteComment = (commentId, reason) => call('delete', `/admin/comments/${commentId}`, { data: { reason } })
export const adminGetReports = (params = {}) => call('get', '/admin/reports', { params })
export const adminResolveReport = (reportId, payload) => call('patch', `/admin/reports/${reportId}`, { data: payload })
export const adminGetInquiries = (params = {}) => call('get', '/admin/inquiries', { params })
export const adminGetInquiry = (inquiryId) => call('get', `/admin/inquiries/${inquiryId}`)
export const adminAnswerInquiry = (inquiryId, payload) => call('patch', `/admin/inquiries/${inquiryId}/answer`, { data: payload })
export const adminGetAuditLogs = (params = {}) => call('get', '/admin/audit-logs', { params })

File: backend/app/routes/features.py
"""Additional user APIs, following the existing /api success/error contract."""
from io import BytesIO

from flask import Blueprint, request, send_file

from app.extensions import db
from app.models.features import SavingGoal, AssetSnapshot, Badge, Post, Comment, Inquiry, Report, Attachment
from app.models.deposits_savings import LedgerTransaction
from app.models.market import MarketTransaction
from app.routes._helpers import require_json_body, success_response
from app.schemas import features as schemas
from app.services.feature_common import endpoint, user_id, get_row, serialize, page, fail, current_user
from app.services import portfolio_service as portfolio, community_service as community, profile_service as profiles
from app.services.projection_service import project
from app.services.account_service import get_account_by_user_id
from app.services.audit_service import audit

features_bp = Blueprint("features", __name__, url_prefix="/api")


def body(schema, partial=False):
    payload = schema().load(require_json_body(), partial=partial)
    if not payload:
        fail("INVALID_REQUEST", "변경할 항목을 입력해 주세요.")
    return payload


@features_bp.get("/goals")
@endpoint()
def goals():
    result = portfolio.dashboard(user_id())
    return success_response(result["goals"])


@features_bp.post("/goals")
@endpoint()
def create_goal():
    return success_response(portfolio.save_goal(user_id(), body(schemas.GoalSchema)), 201)


@features_bp.get("/goals/<int:goal_id>")
@endpoint()
def get_goal(goal_id):
    row = get_row(SavingGoal, goal_id, owner=user_id())
    return success_response(portfolio.goal_data(row, portfolio.valuation(user_id())["total_assets"]))


@features_bp.patch("/goals/<int:goal_id>")
@endpoint()
def update_goal(goal_id):
    return success_response(portfolio.save_goal(user_id(), body(schemas.GoalSchema, partial=True), goal_id))


@features_bp.delete("/goals/<int:goal_id>")
@endpoint()
def delete_goal(goal_id):
    get_account_by_user_id(user_id())
    db.session.delete(get_row(SavingGoal, goal_id, owner=user_id(), lock=True))
    db.session.commit()
    return success_response({})


@features_bp.post("/simulations/free")
@endpoint()
def free_projection():
    return success_response(project(require_json_body()))


@features_bp.post("/goals/<int:goal_id>/simulation")
@endpoint()
def goal_projection(goal_id):
    return success_response(project(require_json_body(), user_id(), goal_id))


@features_bp.get("/dashboard")
@endpoint()
def dashboard():
    return success_response(portfolio.dashboard(user_id()))


@features_bp.get("/dashboard/history")
@endpoint()
def history():
    return success_response(page(AssetSnapshot.query.filter_by(user_id=user_id()).order_by(AssetSnapshot.snapshot_date.desc())))


def ledger_data(row):
    return {**serialize(row), "entries": [serialize(e) for e in row.entries]}


@features_bp.get("/transactions")
@endpoint()
def transactions():
    query = LedgerTransaction.query.filter_by(user_id=user_id())
    kind = request.args.get("transaction_type")
    if kind:
        if len(kind) > 50:
            fail("INVALID_REQUEST", "거래 구분이 너무 깁니다.")
        query = query.filter_by(transaction_type=kind)
    return success_response(page(query.order_by(LedgerTransaction.ledger_transaction_id.desc()), ledger_data))


@features_bp.get("/transactions/<int:transaction_id>")
@endpoint()
def transaction_detail(transaction_id):
    return success_response(ledger_data(get_row(LedgerTransaction, transaction_id, owner=user_id())))


@features_bp.get("/investments/transactions")
@endpoint()
def investment_transactions():
    return success_response(page(MarketTransaction.query.filter_by(user_id=user_id()).order_by(MarketTransaction.market_transaction_id.desc())))


@features_bp.get("/profiles/me")
@endpoint()
def my_profile():
    return success_response(profiles.profile(user_id(), user_id()))


@features_bp.get("/profiles/<int:target_id>")
@endpoint()
def public_profile(target_id):
    return success_response(profiles.profile(target_id, user_id()))


@features_bp.patch("/profiles/me")
@endpoint()
def update_profile():
    return success_response(profiles.update_profile(user_id(), body(schemas.ProfileSchema)))


@features_bp.get("/profiles/me/visibility")
@endpoint()
def visibility():
    return success_response(profiles.visibility(user_id()))


@features_bp.patch("/profiles/me/visibility")
@endpoint()
def update_visibility():
    return success_response(profiles.update_visibility(user_id(), body(schemas.VisibilitySchema)))


@features_bp.get("/badges")
@endpoint()
def badges():
    return success_response([serialize(b) for b in Badge.query.order_by(Badge.badge_id)])


@features_bp.get("/badges/me")
@endpoint()
def my_badges():
    return success_response(portfolio.dashboard(user_id())["badges"])


@features_bp.get("/posts")
@endpoint()
def posts():
    return success_response(community.list_posts(viewer_id=user_id()))


@features_bp.post("/posts")
@endpoint()
def create_post():
    return success_response(community.save_post(user_id(), body(schemas.PostSchema)), 201)


@features_bp.get("/posts/<int:post_id>")
@endpoint()
def post_detail(post_id):
    return success_response(community.post_data(
        community.live_post(post_id),
        viewer_id=user_id(),
    ))


@features_bp.patch("/posts/<int:post_id>")
@endpoint()
def update_post(post_id):
    return success_response(community.save_post(user_id(), body(schemas.PostSchema, partial=True), post_id))


@features_bp.delete("/posts/<int:post_id>")
@endpoint()
def delete_post(post_id):
    community.delete_content("POST", post_id, owner=user_id())
    return success_response({})


@features_bp.get("/posts/<int:post_id>/comments")
@endpoint()
def comments(post_id):
    community.live_post(post_id)
    return success_response(page(Comment.query.filter_by(post_id=post_id, deleted_at=None).order_by(Comment.comment_id), lambda c: serialize(c, exclude=("deleted_at",))))


@features_bp.post("/posts/<int:post_id>/comments")
@endpoint()
def create_comment(post_id):
    return success_response(community.save_comment(user_id(), body(schemas.CommentSchema), post_id=post_id), 201)


@features_bp.patch("/comments/<int:comment_id>")
@endpoint()
def update_comment(comment_id):
    return success_response(community.save_comment(user_id(), body(schemas.CommentSchema), comment_id=comment_id))


@features_bp.delete("/comments/<int:comment_id>")
@endpoint()
def delete_comment(comment_id):
    community.delete_content("COMMENT", comment_id, owner=user_id())
    return success_response({})


@features_bp.put("/posts/<int:post_id>/reaction")
@endpoint()
def reaction(post_id):
    return success_response(community.react(user_id(), post_id, body(schemas.ReactionSchema)["reaction_type"]))


@features_bp.post("/reports")
@endpoint()
def create_report():
    return success_response(community.create_report(user_id(), body(schemas.ReportSchema)), 201)


@features_bp.get("/reports/me")
@endpoint()
def my_reports():
    return success_response(page(Report.query.filter_by(reporter_user_id=user_id()).order_by(Report.report_id.desc())))


@features_bp.post("/inquiries")
@endpoint()
def create_inquiry():
    return success_response(community.create_inquiry(user_id(), body(schemas.InquirySchema)), 201)


@features_bp.get("/inquiries")
@endpoint()
def inquiries():
    return success_response(page(Inquiry.query.filter_by(user_id=user_id()).order_by(Inquiry.inquiry_id.desc()), community.inquiry_data))


@features_bp.get("/inquiries/<int:inquiry_id>")
@endpoint()
def inquiry_detail(inquiry_id):
    return success_response(community.inquiry_data(get_row(Inquiry, inquiry_id, owner=user_id())))


@features_bp.post("/posts/<int:post_id>/images")
@endpoint()
def post_images(post_id):
    return success_response(community.upload(user_id(), "POST", post_id, request.files.getlist("images")), 201)


@features_bp.post("/inquiries/<int:inquiry_id>/images")
@endpoint()
def inquiry_images(inquiry_id):
    return success_response(community.upload(user_id(), "INQUIRY", inquiry_id, request.files.getlist("images")), 201)


@features_bp.get("/attachments/<string:attachment_id>")
@endpoint()
def attachment(attachment_id):
    row = get_row(Attachment, attachment_id)
    if row.post_id:
        community.live_post(row.post_id)
    elif row.user_id != user_id() and current_user().role != "ADMIN":
        fail("NOT_FOUND", "첨부파일을 찾을 수 없습니다.", 404)
    response = send_file(BytesIO(row.data), mimetype=row.mime_type, download_name="image", max_age=0)
    response.headers["X-Content-Type-Options"] = "nosniff"
    response.headers["Cache-Control"] = "private, no-store"
    return response


@features_bp.delete("/attachments/<string:attachment_id>")
@endpoint()
def delete_attachment(attachment_id):
    row = get_row(Attachment, attachment_id, owner=user_id(), lock=True)
    audit("IMAGE_DELETE", "posts" if row.post_id else "inquiries", row.post_id or row.inquiry_id)
    db.session.delete(row)
    db.session.commit()
    return success_response({})
